renormalizer/utils/basis.py

```diff
@@ -46,113 +47,90 @@ def op_mat(self, op):
 
-            mat = np.eye(self.nbas) * self.x0
-            # <m|y|n> = sqrt(1/2w) <m|b^dagger b |n> = sqrt(n+1) delta(m, n+1) +
-            # sqrt(n) delta(m, n-1)
-            for iket in range(self.nbas-1):
-                mat[iket+1, iket] += np.sqrt(iket+1) * np.sqrt(0.5/self.omega)
-            for iket in range(1, self.nbas):
-                mat[iket-1, iket] += np.sqrt(iket) * np.sqrt(0.5/self.omega)
+        elif op_symbol == "x":
+            # define x-x0 = y or x = y+x0, return x
+            # <m|y|n> = sqrt(1/2w) <m| b^\dagger + b |n>
+            mat = np.sqrt(0.5/self.omega) * self.op_mat(r"b^\dagger + b") + np.eye(self.nbas) * self.x0
 
         elif op_symbol == "x^2":
+            # can't do things like the commented code below due to numeric error around highest quantum number
+            # x_mat = self.op_mat("x")
+            # mat = x_mat @ x_mat
+
+            # x^2 = x0^2 + 2 x0 * y + y^2
             # x0^2
             mat = np.eye(self.nbas) * self.x0**2
-            
-            # 2 x0 * y 
-            for iket in range(self.nbas-1):
-                mat[iket+1, iket] += np.sqrt(iket+1) * np.sqrt(0.5/self.omega) * 2. * self.x0
-            for iket in range(1, self.nbas):
-                mat[iket-1, iket] += np.sqrt(iket) * np.sqrt(0.5/self.omega) * 2. * self.x0
-            
-            #  y^2: b^dagger b + b b^\dagger = (n+1+n) delta(m,n)
-            for iket in range(self.nbas):
-                mat[iket, iket] += (2*iket+1) * 0.5/self.omega
-            
-            #  y^2: bb = sqrt(n*(n-1)) delta(m,n-2)
-            for iket in range(2, self.nbas):
-                mat[iket-2, iket] += np.sqrt(iket*(iket-1)) * 0.5/self.omega
 
-            #  y^2: b^\dagger b^\dagger = sqrt((n+2)*(n+1)) delta(m,n+2)
-            for iket in range(0, self.nbas-2):
-                mat[iket+2, iket] += np.sqrt((iket+2)*(iket+1)) * 0.5/self.omega
-        
+            # 2 x0 * y
+            mat += 2 * self.x0 * np.sqrt(0.5/self.omega) * self.op_mat(r"b^\dagger + b")
+
+            #  y^2: 1/2w * (b^\dagger b^\dagger + b^dagger b + b b^\dagger + bb)
+            mat += 0.5/self.omega * (self.op_mat(r"b^\dagger b^\dagger")
+                                     + self.op_mat(r"b^\dagger b")
+                                     + self.op_mat(r"b b^\dagger")
+                                     + self.op_mat(r"b b")
+                                     )
+
         elif op_symbol == "p":
-            #<m|p|n> = -i sqrt(w/2) sqrt(n) delta(m,n-1) - sqrt(n+1) delta(m,n+1)
-            mat = np.zeros((self.nbas, self.nbas), dtype=np.complex128)
-            
-            for iket in range(1, self.nbas):
-                mat[iket-1, iket] += -1.0j * np.sqrt(0.5*self.omega) * np.sqrt(iket) 
-            for iket in range(self.nbas-1):
-                mat[iket+1, iket] -= -1.0j * np.sqrt(0.5*self.omega) * np.sqrt(iket+1)
+            # <m|p|n> = -i sqrt(w/2) <m| b^\dagger - b |n>
```

Comment:
I think the second quantization of p is wrong. 
-i sqrt(w/2) <m| b - b^\dagger |n>

Reply:
Indeed. This could be fixed in future commits.